### PR TITLE
Generated source code from template

### PR DESCRIPTION
**Pull request by vidhutsingh**
Feature : GoLangSDK 
Developed  templates for object, objectSet, enum, objectSetFactory for generation SDK client packages. 
1. object - 
	1.Handled complex data type, object , array of objects and array of 	complex datatypes .
 	2.Generated all the nimos objects.
2. objectSet -
    1. Generated all the objectSets method for the objects.  
   3. ObjectSetFactory 
	1. This contains the GetObjectSet methods (kind of constructor ) for 
	    all the objectSets i.e VolumeObjectSet(). 
   4. Enum 
	Creates a template for enum types and it generated all the nimos enum

Converted all the object’s member to pointer type as it was causing issue during communication to Array firmware. Our object contains both request and response members in the same object (structure) thus firmware was throwing error “SM_unexpected_arg” since it was not expecting response member during request. By using pointers types this is handled as nil pointers are dropped during API communication. 

Made changes to volume_service_test suite 

Test :
Go test . 
Able to perform all the basic operation on Array.

### CODE REVIEW

File: pkg/client/access_control_record_object_set.go [hunk 1]
```diff
@@ -1,70 +1,76 @@
...
-
-	for i := 0; i < values.Len(); i++ {
-		value := &model.AccessControlRecord{}
-		jsonutil.Decode(values.Index(i).Interface(), value)
-		results[i] = value
-	}
-
-	return results
-}
+package client
```
Comment (rgcostea) [inline pkg/client/access_control_record_object_set.go:3]: We lost the copyright notice

File: pkg/client/access_control_record_object_set.go [hunk 2]
```diff
@@ -1,70 +1,76 @@
...
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/client/v1/model"
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/util"
+)
+
+/**
+ * Manage access control records for volumes.
+ *
+ */
+const (
+    accessControlRecordPath = "access_control_records"
```
Comment (rgcostea) [inline pkg/client/access_control_record_object_set.go:16]: It looks like tabs got replaced with space throughout.  golang favors tabs.  We need to fix that.

File: pkg/client/v1/model/NsALUA.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Copyright 2017 Hewlett Packard Enterprise Development LP
+ */
```
Comment (rgcostea) [inline pkg/client/v1/model/NsALUA.go]: The year is off and it doesn't follow the typical comment pattern for copyright

File: pkg/client/v1/model/NsALUA.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+package model
+
+
+/**
+ * <p>Golang package for NsALUA.</p>
+ */
+
+type NsALUA string
```
Comment (rgcostea) [inline pkg/client/v1/model/NsALUA.go]: The naming pattern is off for all of these ns types.  They should follow the rest of the convention. 

ns_alua.go, ns_access_apply_to.go, etc.

File: pkg/client/v1/model/access_control_record.go [hunk 5]
```diff
@@ -3,39 +3,121 @@
  */
 
 package model
-//package nimblestorage/v1/AccessControlRecord
 
 
-// AccessControlRecord :
+
+// AccessControlRecord - Manage access control records for volumes.
+// Export AccessControlRecordFields for advance operations like search filter etc.
```
Comment (rgcostea) [inline pkg/client/v1/model/access_control_record.go]: Missing copyright

File: pkg/client/v1/model/active_directory_membership.go [hunk 6]
```diff
@@ -3,29 +3,77 @@
...
+func init(){
+	IDfield:= "id"
+	Descriptionfield:= "description"
+	Namefield:= "name"
+	Netbiosfield:= "netbios"
+	ServerListfield:= "server_list"
+	ComputerNamefield:= "computer_name"
+	OrganizationalUnitfield:= "organizational_unit"
+	Userfield:= "user"
+	Passwordfield:= "password"
```
Comment (rgcostea) [inline pkg/client/v1/model/active_directory_membership.go:19]: Why do we need to declare these here as <parameterName>field when we simply use them down below in the <Object>Fields initializer?

Just set the value directly on the declaration below and we eliminate the need for all these extra lines.

File: pkg/client/v1/model/volume.go [hunk 7]
```diff
@@ -4,182 +4,478 @@
...
+    // Name of performance policy.
+    
+ 	PerfpolicyName *string `json:"perfpolicy_name,omitempty"`
+   
+    // Identifier of the performance policy. After creating a volume, performance policy for the volume can only be changed to another performance policy with same block size.
+    
+ 	PerfpolicyID *string `json:"perfpolicy_id,omitempty"`
+   
+    // Amount of space to reserve for this volume as a percentage of volume size.
+
```
Comment (rgcostea) [inline pkg/client/v1/model/volume.go]: Let's drop these extra lines between comment and name.

File: pkg/client/v1/model/volume.go [hunk 8]
```diff
@@ -4,182 +4,478 @@
...
+    
+ 	PerfpolicyID *string `json:"perfpolicy_id,omitempty"`
+   
+    // Amount of space to reserve for this volume as a percentage of volume size.
+    
+   	Reserve *int64 `json:"reserve,omitempty"`
+   
+    // This attribute is deprecated. Alert threshold for the volume's mapped usage, expressed as a percentage of the volume's size. When the volume's mapped usage exceeds warn_level, the array issues an alert. If this option is not specified, array default volume warn level setting is used to decide the warning level for this volume.
+    
+   	WarnLevel *int64 `json:"warn_level,omitempty"`
```
Comment (rgcostea) [inline pkg/client/v1/model/volume.go]: Note that golang fmt requires the name of the field to be the first part of the comment...

// WarnLevel - <something>
WarnLevel *int64

File: pkg/client/v1/model/initiator_group.go [hunk 9]
```diff
@@ -3,33 +3,109 @@
...
+    
+   	CreationTime *int64 `json:"creation_time,omitempty"`
+   
+    // Time when this initiator group was last modified.
+    
+   	LastModified *int64 `json:"last_modified,omitempty"`
+   
+    // Flag to allow modifying VP created initiator groups. When set to true, user can add this initiator to a VP created initiator group.
+    
+ 	VpOverrIDe *bool `json:"vp_override,omitempty"`
```
Comment (rgcostea) [inline pkg/client/v1/model/initiator_group.go]: It looks like the generator is NOT reading the hidden field attribute.  This field should not be exposed publicly.

File: pkg/client/v1/model/initiator_group.go [hunk 10]
```diff
@@ -3,33 +3,109 @@
...
+   
+    // Application identifier of initiator group.
+    
+ 	AppUuID *string `json:"app_uuid,omitempty"`
```
Comment (rgcostea) [inline pkg/client/v1/model/initiator_group.go]: Also looks like the generator needs some help here.

AppUuID is not correct

**Review by rgcostea — CHANGES_REQUESTED** [review]
Please address comments.

File: pkg/client/v1/model/active_directory_membership.go (hunk 6 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/active_directory_membership.go:19]: Object members are pointers.

Comment (rgcostea) [reply pkg/client/v1/model/active_directory_membership.go:19]: Ya.  That's right.  And there's no clean one-liner to do this in golang.  Oh well.  Thanks.

File: pkg/client/v1/model/initiator_group.go (hunk 10 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/initiator_group.go]: We have app_uuid under initiator_group which is converted to camel case and id is replaced with ID intentionally. So what exactly is the issue here ?

File: pkg/client/access_control_record_object_set.go (hunk 1 above)
Comment (vidhutsingh) [reply pkg/client/access_control_record_object_set.go:3]: Thanks. Will fix it .

File: pkg/client/access_control_record_object_set.go (hunk 2 above)
Comment (vidhutsingh) [reply pkg/client/access_control_record_object_set.go:16]: Need to handle it in the template . will fix it.

File: pkg/client/v1/model/NsALUA.go (hunk 3 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/NsALUA.go]: Will fix it . thanks

File: pkg/client/v1/model/NsALUA.go (hunk 4 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/NsALUA.go]: Sure, will fix it .

File: pkg/client/v1/model/access_control_record.go (hunk 5 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/access_control_record.go]: Will fix it.

File: pkg/client/v1/model/volume.go (hunk 7 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/volume.go]: Sure, will fix it.

File: pkg/client/v1/model/volume.go (hunk 8 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/volume.go]: Sure will fix it. thanks

File: pkg/client/v1/model/initiator_group.go (hunk 9 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/initiator_group.go]: Need to investigate. Right now I've not made any changes related to internal vs external.

**Comment by vidhutsingh** [comment]
> Please address comments.

Thanks for review comments @rgcostea !

File: pkg/client/v1/model/initiator_group.go (hunk 10 above)
Comment (rgcostea) [reply pkg/client/v1/model/initiator_group.go]: The issue is that id is replaced with ID intentionally.  Look at these 2 examples:
VpOverrIDe should be VpOverride
AppUuID should be AppUuid

We need to handle where id might be used within any field and not just blindly convert it to ID.

Comment (vidhutsingh) [reply pkg/client/v1/model/initiator_group.go]: Ok. will fix it.

**Comment by vidhutsingh** [comment]
@rgcostea , please check review changes. thanks !

File: pkg/client/access_control_record_object_set.go [hunk 11]
```diff
@@ -1,70 +1,78 @@
...
+
+	"github.com/hpe-storage/common-host-libs/jsonutil"
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/client/v1/model"
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/util"
+)
+
+/**
+ * Manage access control records for volumes.
+ *
+ */
```
Comment (rgcostea) [inline pkg/client/access_control_record_object_set.go]: Use golang comment pattern (//)

File: pkg/client/access_control_record_object_set.go [hunk 12]
```diff
@@ -1,70 +1,78 @@
...
+const (
+    accessControlRecordPath = "access_control_records"
+)
+
+/**
+ * AccessControlRecordObjectSet
+*/
```
Comment (rgcostea) [inline pkg/client/access_control_record_object_set.go]: here too

File: pkg/client/v1/model/initiator_group.go [hunk 13]
```diff
@@ -1,35 +1,70 @@
...
+	// IscsiInitiators - List of iSCSI initiators. When create/update iscsi_initiators, either iqn or ip_address is always required with label.
+   	IscsiInitiators []*NsISCSIInitiator `json:"iscsi_initiators,omitempty"`
+	// FcInitiators - List of FC initiators. When create/update fc_initiators, wwpn is required.
+   	FcInitiators []*NsFCInitiator `json:"fc_initiators,omitempty"`
+	// CreationTime - Time when this initiator group was created.
+   	CreationTime *int64 `json:"creation_time,omitempty"`
+	// LastModified - Time when this initiator group was last modified.
+   	LastModified *int64 `json:"last_modified,omitempty"`
+	// VpOverride - Flag to allow modifying VP created initiator groups. When set to true, user can add this initiator to a VP created initiator group.
+ 	VpOverride *bool `json:"vp_override,omitempty"`
```
Comment (rgcostea) [inline pkg/client/v1/model/initiator_group.go]: Hidden parameters need to be removed before you push.

File: pkg/client/v1/model/initiator_group.go [hunk 14]
```diff
@@ -1,35 +1,70 @@
...
+	AppUuidfield:= "app_uuid"
+		
+	InitiatorGroupFields= &InitiatorGroup{
+	ID: &IDfield,
+	Name: &Namefield,
+	FullName: &FullNamefield,
+	SearchName: &SearchNamefield,
+	Description: &Descriptionfield,
+	HostType: &HostTypefield,
+	AppUuid: &AppUuidfield,
```
Comment (rgcostea) [inline pkg/client/v1/model/initiator_group.go]: indent

Comment (vidhutsingh) [reply pkg/client/v1/model/initiator_group.go:26]: I tried to indent but freemarker doesnt have rich syntaxes to do that.

Comment (vidhutsingh) [reply pkg/client/v1/model/initiator_group.go]: Done.

File: pkg/client/v1/model/initiator_group.go (hunk 13 above)
Comment (vidhutsingh) [reply pkg/client/v1/model/initiator_group.go]: Thsi is handled.

File: pkg/client/access_control_record_object_set.go (hunk 11 above)
Comment (vidhutsingh) [reply pkg/client/access_control_record_object_set.go]: ok. thx.

**Review by rgcostea — APPROVED** [review]
lgtm.  Let's get @suneeth51 to look at this too.

File: pkg/client/active_directory_membership_object_set.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	"reflect"
+
+	"github.com/hpe-storage/common-host-libs/jsonutil"
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/client/v1/model"
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/util"
+)
+
+
+// Manages the storage array&#39;s membership with the Active Directory.
```
Comment (suneeth51) [inline pkg/client/active_directory_membership_object_set.go]: nit: Need to remove those unwanted characters in  `array&#39;s`
`Manages the storage array membership with the Active Directory`

File: pkg/client/fibre_channel_session_object_set.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+// DeleteObject deletes the FibreChannelSession object with the specified ID
+func (objectSet *FibreChannelSessionObjectSet) DeleteObject(id string) error {
+	return objectSet.Client.Delete(fibreChannelSessionPath, id)
+}
+
+// GetObject returns a FibreChannelSession object with the given ID
+func (objectSet *FibreChannelSessionObjectSet) GetObject(id string) (*model.FibreChannelSession, error) {
+	response, err := objectSet.Client.Get(fibreChannelSessionPath, id, model.FibreChannelSession{})
+	if response == nil {
+		return nil, err
```
Comment (suneeth51) [inline pkg/client/fibre_channel_session_object_set.go:45]: nit: I think the tabs needs to be fixed throughout the code.

File: pkg/client/group_mgmt_client.go [hunk 17]
```diff
@@ -75,13 +75,14 @@ func (client *GroupMgmtClient) EnableDebug() {
 
 func (client *GroupMgmtClient) login(username, password string) (string, error) {
 	// Construct Payload
+	appName := "Go client"
```
Comment (suneeth51) [inline pkg/client/group_mgmt_client.go]: We need to append golang SDK version info as well as part of Go client login.

File: pkg/client/performance_policy_object_set.go [hunk 18]
```diff
@@ -1,66 +1,75 @@
...
+import (
+	"reflect"
+
+	"github.com/hpe-storage/common-host-libs/jsonutil"
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/client/v1/model"
+	"github.hpe.com/nimble-dcs/golang-sdk/pkg/util"
+)
+
+
+// Manage performance policies. A performance policy is a set of optimizations including block size, compression, and caching, to ensure that the volume&#39;s performance is the best configuration for its intended use like databases or log files. By default, a volume uses the \\&quot;default\\&quot; performance policy, which is set to use 4096 byte blocks with full compression and caching enabled. For replicated volumes, the same performance policy must exist on each replication partner.
```
Comment (suneeth51) [inline pkg/client/performance_policy_object_set.go]: nit: `compression and caching to ensure`

**Review by suneeth51 — APPROVED** [review]
lgtm with few minor comments.

@vidhutsingh We have generated all endpoints (including internal) here. We want to skip those internal endpoints/attributes/actions exposing to Public github. We need to work on fixing the Templates to generate accordingly.

**Comment by vidhutsingh** [comment]
> lgtm with few minor comments.
> 
> @vidhutsingh We have generated all endpoints (including internal) here. We want to skip those internal endpoints/attributes/actions exposing to Public github. We need to work on fixing the Templates to generate accordingly.

Yes this has to be fixed as external vs internals endpoints are not handled. are you suggesting with this PR or in subsequent PRs ?

File: pkg/client/fibre_channel_session_object_set.go (hunk 16 above)
Comment (vidhutsingh) [reply pkg/client/fibre_channel_session_object_set.go:45]: This looks okay in the vs code. Looks like browser showing as if there is more space than a tab.

File: pkg/client/group_mgmt_client.go (hunk 17 above)
Comment (vidhutsingh) [reply pkg/client/group_mgmt_client.go]: Ok, let me use hard coded value for now . As the we come near to sdk release we will append the version accordingly. thanks.

File: pkg/client/active_directory_membership_object_set.go (hunk 15 above)
Comment (vidhutsingh) [reply pkg/client/active_directory_membership_object_set.go]: Sure, will fix it .

**Comment by rgcostea** [comment]
> Yes this has to be fixed as external vs internals endpoints are not handled. are you suggesting with this PR or in subsequent PRs ?

Let's do it in this PR please.

**Pull request merged**